libass/ass_render.c

```diff
@@ -3392,25 +3544,73 @@ ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
     }
 
     // render events separately
-    int cnt = 0;
+    int sent = 0, cnt = 0;
     for (int i = 0; i < track->n_events; i++) {
         ASS_Event *event = track->events + i;
         if ((event->Start <= now)
             && (now < (event->Start + event->Duration))) {
-            if (cnt >= priv->eimg_size) {
+            if (sent >= priv->eimg_size) {
                 priv->eimg_size += 100;
                 priv->eimg =
                     realloc(priv->eimg,
                             priv->eimg_size * sizeof(EventImages));
             }
-            if (ass_render_event(&priv->state, event, priv->eimg + cnt))
+            priv->eimg[sent++].event = event;
+        }
+    }
+
+#if ENABLE_THREADS
+    if (priv->settings.threads == 0)
+        priv->settings.threads = default_threads();
+
+    if (!priv->n_threads && !priv->thread_start_failed && priv->settings.threads > 1) {
+        if (!(priv->threads = calloc(priv->settings.threads, sizeof(ASSThread)))) {
+            ass_msg(priv->library, MSGL_ERR, "Allocation failure");
+            return NULL;
+        }
+
+        for (priv->n_threads = 0; priv->n_threads < priv->settings.threads; priv->n_threads++) {
+            if (!assi_thread_create(&priv->threads[priv->n_threads], run_thread, priv)) {
+                assi_mutex_lock(&priv->mutex);
+                ass_msg(priv->library, MSGL_WARN, "Thread startup failure");
+                priv->thread_start_failed = 1;
+                assi_mutex_unlock(&priv->mutex);
+                goto thread_fail;
+            }
+        }
+
+        assi_mutex_lock(&priv->mutex);
+        while (priv->started_threads < priv->n_threads && !priv->thread_start_failed)
+            assi_cond_wait(&priv->main_cond, &priv->mutex);
+        assi_mutex_unlock(&priv->mutex);
+    }
+
+thread_fail:
+    if (priv->n_threads > 0 && !priv->thread_start_failed && sent > 1) {
+        assi_mutex_lock(&priv->mutex);
+        priv->next_eimg = priv->got_eimgs = 0;
+        priv->processing_eimgs = priv->sent_eimgs = sent;
+
+        assi_cond_broadcast(&priv->pool_cond);
+
+        while (priv->processing_eimgs)
+            assi_cond_wait(&priv->main_cond, &priv->mutex);
+
+        cnt = priv->got_eimgs;
+
+        assi_mutex_unlock(&priv->mutex);
+    } else
+#endif
+    {
```

Comment:
`{` isn't really needed here.

Reply:
I prefer to use a block when wrapping an inner control-flow statement like this.